dhall-nix/src/Dhall/Nix.hs

```diff
@@ -215,10 +220,77 @@ Right x: y: x + y
     the expression to `dhallToNix`
 -}
 dhallToNix :: Expr s Void -> Either CompileError (Fix NExprF)
-dhallToNix e = loop (Dhall.Core.normalize e)
+dhallToNix e =
+    loop (rewriteShadowed (Dhall.Core.normalize e))
   where
     untranslatable = Fix (NSet NNonRecursive [])
 
+    -- This is an intermediate utility used to remove all occurrences of
+    -- shadowing (since Nix does not support references to shadowed variables)
+    --
+    -- This finds how many bound variables of the same name that we need to
+    -- descend past to reach the "deepest" reference to the current bound
+    -- variable.  In other words, the result is the "depth" of the deepest
+    -- reference.
+    --
+    -- If `Nothing` then the current bound variable doesn't need to be renamed.
+    -- If any other number, then rename the variable to include the maximum
+    -- depth.
+    maximumDepth :: Var -> Expr s Void -> Maybe Int
+    maximumDepth v@(V x n) (Lam x' a b)
+        | x == x' =
+            max (maximumDepth v a) (fmap (+ 1) (maximumDepth (V x (n + 1)) b))
+    maximumDepth v@(V x n) (Pi x' a b)
+        | x == x' =
+            max (maximumDepth v a) (fmap (+ 1) (maximumDepth (V x (n + 1)) b))
+    maximumDepth (V x n) (Let (Binding { variable = x' }) a)
+        | x == x' = fmap (+ 1) (maximumDepth (V x (n + 1)) a)
+    maximumDepth v (Var v')
+        | v == v' = Just 0
+    maximumDepth v expression =
+        foldr max Nothing
+            (map
+                (maximumDepth v)
+                (toListOf Dhall.Core.subExpressions expression)
+            )
```

Comment:
So for each lambda it has to recurse into the whole subexpression to search for similarly named expressions. What is the complexity of this? Superficially it looks like it’s O(n^2) at least, maybe more.

Reply:
My initial idea was to add new variables to an ambient stack, and whenever a new variable is introduced you would check against the stack whether that variable name was already used.
That would be a linear lookup in the current subexpression depth, so more like O(log(n)).

Reply:
@Profpatsch: The main reason I didn't do it that way was because I didn't want to rename variables if it wasn't necessary.  For example, for the expression:

```dhall
λ(x : Bool) → λ(x : Bool) → x
```

... no renaming is necessary even though the outer `x` is shadowed.

Reply:
However, if performance of this ever does become an issue I'll switch to the simpler algorithm that unconditionally renames things when shadowing occurs

Reply:
I’m fine with unconditional renaming, since most of the time the variable will be named `_` anyway because of alpha-renaming …

Reply:
Yeah, that may be the case right now, but one of my queued changes to the language is to add a variation on integrity checks that does not α-normalize imports (e.g. for caching purposes, rather than security protection).  For more details, see: https://github.com/dhall-lang/dhall-lang/issues/733

If we fix that then the variables would have useful names that I'd like `dhall-to-nix` to preserve